Applications/Embedding/jni/main.cpp

```diff
@@ -0,0 +1,282 @@
+}
+
+/**
+ * @brief     get Data as much as batch size
+ * @param[out] outVec feature data
+ * @param[out] outLabel label data
+ * @param[out] last end of data
+ * @param[in] user_data user data
+ * @retval int 0 if there is no error
+ */
+int getBatch_train(float **outVec, float **outLabel, bool *last,
+                   void *user_data) {
+  std::ifstream dataFile(data_file);
+  unsigned int data_size = total_train_data_size;
+  unsigned int count = 0;
+
+  if (data_size - train_count < batch_size) {
+    *last = true;
+    train_count = 0;
+    return 0;
+  }
+
+  for (unsigned int i = train_count; i < train_count + batch_size; ++i) {
+
+    std::vector<float> o;
+    std::vector<float> l;
+    o.resize(feature_size);
+    l.resize(1);
+
+    if (!getData(dataFile, o, l, i)) {
+      return -1;
+    };
+
+    for (unsigned int j = 0; j < feature_size; ++j)
+      outVec[0][count * feature_size + j] = o[j];
+    outLabel[0][count] = l[0];
+
+    count++;
+  }
+
+  dataFile.close();
+  *last = false;
+  train_count += batch_size;
+  return 0;
+}
+
+/**
+ * @brief     create NN
+ *            back propagation of NN
+ * @param[in]  arg 1 : train / inference
+ * @param[in]  arg 2 : configuration file path
+ * @param[in]  arg 3 : resource path (dataset.txt or testset.txt)
+ */
+int main(int argc, char *argv[]) {
+  if (argc < 4) {
+    std::cout << "./Embedding train (| inference) Config.ini data.txt\n";
+    exit(1);
+  }
+
+  const std::vector<std::string> args(argv + 1, argv + argc);
+  std::string config = args[1];
+  data_file = args[2];
+
+  if (!args[0].compare("train"))
+    training = true;
+
+  srand(time(NULL));
+
+  std::shared_ptr<ml::train::Dataset> dataset;
+  try {
+    dataset = createDataset(ml::train::DatasetType::GENERATOR);
+    dataset->setGeneratorFunc(ml::train::DatasetDataType::DATA_TRAIN,
+                              getBatch_train);
+    dataset->setGeneratorFunc(ml::train::DatasetDataType::DATA_VAL,
+                              getBatch_train);
+  } catch (std::exception &e) {
+    std::cerr << "Error creating dataset" << e.what() << std::endl;
+    return 1;
+  }
+
+  /**
+   * @brief     Create NN
+   */
+  std::vector<std::vector<float>> inputVector, outputVector;
+  nntrainer::NeuralNetwork NN;
+  std::shared_ptr<nntrainer::Layer> layer;
+  std::shared_ptr<nntrainer::Layer> layer_fc;
+  std::string name = "embedding";
+  std::string fc_name = "outputlayer";
+  nntrainer::Tensor weight;
+  nntrainer::Tensor weight_fc;
+  /**
+   * @brief     Initialize NN with configuration file path
+   */
+
+  try {
+    NN.loadFromConfig(config);
+    NN.compile();
+    NN.initialize();
+    NN.readModel();
+
+    nntrainer::Tensor weight_in(1, 1, 15, 8);
+    loadFile("embedding_weight_input.in", weight_in);
+
+    weight_in.print(std::cout);
+
+    NN.getLayer(name.c_str(), &layer);
+    NN.getLayer(fc_name.c_str(), &layer_fc);
+
+    weight = layer->getWeights()[0].getVariable();
+    weight.print(std::cout);
```

Comment:
maybe initialize the weight with weight_in or is it just for comparing the printed values?

Reply:
just comparing. This code is not compareing the output with tf result at this time. It will come up soon after more test.

Reply:
yes, let's add some tests which match the embedding output.